### ISSUE TITLE
Breakage with Tor Browser (observed from 1.0.8.6 to 1.0.10.2+)

### ISSUE THREAD

**Issue by nym-zone** [nym-zone/block_cloudflare_mitm_fx#14]
Earlier, I installed 1.0.8.6 on Tor Browser “7.0.11 (based on Mozilla Firefox 52.5.2) (64-bit)”.  The whitelisting and options UI did not work.

I tried to open the Browser Console to figure out what was happening; but the whole browser froze, spinning my CPU, and I had to kill it.  Thus, I could not get any useful debug information; and I have not had time to investigate further.  Also I’ve not had time to test adequately against older versions, to see if it be a regression.  Apologies for the substance-starved bug report.

@whatsusername, I know that Tor Browser dislikes localstorage in its default configuration.  Could that be the cause?  I know that sites aren’t allowed localstorage; but I have no idea if/how that impacts add-ons.

**Comment by ghost** [comment]
Confirmed with Ff 57.03 x64 on Win 10 x64 b16299, except for the browser freeze when opening console.

`Add Exception` opens the pop up to either `confirm` or `cancel`, but clicking `ok` does _not_ add the exception.

Clicking `View Exceptions` does not do not anything.

**Comment by whatsusername** [comment]
FYI, I always test my add-ons with Firefox 57 and TBB.

@n8v8R Are you using RequestPolicy or something that block connection? If so, please allow connection to 0.0.0.0. Some other add-on of yours are blocking it.

**Comment by whatsusername** [comment]
RequestPolicy alternative ---> https://addons.mozilla.org/en-US/firefox/addon/3p-request-blocker/

**Comment by ghost** [comment]
Yes indeed https://addons.mozilla.org/en-GB/firefox/addon/policy-control/

If active the console is showing `Content Security Policy: The page's settings blocked the loading of a resource at https://0.0.0.0/cfmitm_addon`

With said extension there is no option to unblock `resource` but one of those html elements instead:
✓ Font (@font-face)
✓ Image
✓ Media (audio and video)
✓ Object (plugin-handled content)
✓ Script (JavaScript)
✓ Stylesheet (CSS)
✓ Sub Document (iframes and frames)
✓ XMLHttpRequest
✓ Plugins
✓ WebSocket
✓ Ping
✓ Beacon

Which one to unblock?
0.0.0.0 is it like localhost?

Now having Policy Control disabled the bug is still present. The console output reading `TypeError: proxy must report the same value for the non-writable, non-configurable property '"toDataURL"'`

**Comment by whatsusername** [comment]
Hint: If you're using my [3P blocker](https://addons.mozilla.org/en-US/firefox/addon/3p-request-blocker/) , it'll allow my add-on's connection automatically.
I use my add-ons and NoScript, uBlock Origin, HTTPS everywhere, and nothing else.
On TBB, I use TBB's default add-ons and my 3p blocker only.

> Which one to unblock?

Uh... it's just using .src='' to load the URL, so try "Sub Document".

> 0.0.0.0 is it like localhost?

FYI, localhost will connect to your PC. But 0.0.0.0 is a blackhole. No connection to yourself, it just drop itself.

**Comment by whatsusername** [comment]
> Now having Policy Control disabled the bug is still present.
> The console output reading TypeError: proxy must report the same value for the non-writable,non-configurable property '"toDataURL"'

I never saw anything like that... are you using a proxy? It's fine(I'm using it too), but this smells another add-on conflict.

**Comment by whatsusername** [comment]
> Content Security Policy: The page's settings blocked the loading of a resource

known issue from v1.0.0. some websites use "object-src: none" or something.

**Comment by ghost** [comment]
Meantime upgraded 1.0.9.1

Clicking `View Exceptions` for the first time open the add-on page alright, clicking it again does nothing and the console showing

```
TypeError: can't redefine non-configurable property "getElementsByName"
ObjectOrReflectDefineProperty self-hosted:4079:16 defineProperty self-hosted:4084:5
```

**Comment by ghost** [comment]
`Add Exception` appears to be working now with 1.0.9.1, even with Policy Control enabled and 0.0.0.0 not whitelisted

**Comment by nym-zone** [comment]
I tested with stock Tor Browser (in an ephemeral setup).  Almost all Tor Browser settings were default; the exceptions are such matters as raising the Security Slider (actually, I preset a `prefs.js` entry for that).  No extensions were installed, other than those which are bundled with Tor Browser.

This is the Linux distribution of Tor Browser.

**Comment by whatsusername** [comment]
@nym-zone Delete this add-on from the browser, close it, then add it again. Should work.

@n8v8R please try v1.0.9.2. I replace a button action to link, so CORS or other add-on won't bother you anymore.

**Comment by ghost** [comment]
perfect, works like a charm.

Though not part of this bug I was wondering whether with the subdomains it might confuse new users of this extension having to click several times (n = number of subdomains) until the site is finally loading. Might be worth considering to highlight/bold on the blocking site. That is just a thought since during first exposure I kept on wondering why the block page kept on reappearing.

**Comment by whatsusername** [comment]
> considering to highlight/bold on the blocking site

I'll improve the warning page's UI.

**Comment by whatsusername** [comment]
@nym-zone I think reinstall works (delete add-on, close browser, open browser, install).
The core itself is not changed and I can't reproduce your bug.

**Comment by nym-zone** [comment]
This has been reported by others, as of 1.0.10.2 with Tor Browser:

>I also don’t seem to be able to save preferences.  No checked boxes or added “FQDN”s are retained upon clicking save.  Are the other MITM detections functional?

Good question, that last.  I think no, but I’m not sure.

I am also still experiencing this problem (although not the freeze/CPU-spin I did in 1.0.8.6).
